src/ngraph/runtime/cpu/builder/mlir_cpu_compiled_kernel.cpp

```diff
@@ -65,14 +65,26 @@ namespace ngraph
                     {
                         ptr_args.push_back(ctx->buffer_data[buffer_index]);
                     }
-
                     // Compile nodes within the CompiledKernel op.
-                    auto* compiled_kernel = static_cast<const CompiledKernel*>(node);
+                    CompiledKernel* compiled_kernel =
+                        static_cast<CompiledKernel*>(const_cast<Node*>(node));
+                    bool is_module_ready = true;
+                    auto it = ctx->mlir_compilers.find(compiled_kernel);
+
+                    if (it == ctx->mlir_compilers.end())
+                    {
+                        // create a new compiler for the CK
+                        ctx->mlir_compilers.emplace(compiled_kernel, compiled_kernel);
+                        is_module_ready = false;
+                    }
 
-                    MLIRCompiler mlir_compiler(compiled_kernel, ptr_args);
-                    // TODO: Decouple 'compile' and 'run' APIs. We want to be able to run the same
-                    // jitted code on different arguments.
-                    mlir_compiler.compile_and_run();
+                    MLIRCompiler& mlir_compiler = ctx->mlir_compilers.find(compiled_kernel)->second;
+                    if (!is_module_ready)
+                    {
+                        mlir_compiler.compile();
+                    }
+                    mlir_compiler.set_args(&ptr_args);
+                    mlir_compiler.run();
```

Comment:
I think it would be better if we pass `ptr_args` as reference in `run`. We won't forget to set the args before running and we would avoid dealing with a nullptr most of the time since `m_external_args` will only be valid when we run. Does it make sense?